File: packages/android-client/MobileWhepClient/src/main/java/com/mobilewhep/client/WhipClient.kt
package com.mobilewhep.client

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.util.DisplayMetrics
import android.util.Log
import android.view.WindowManager
import androidx.core.content.ContextCompat
import com.mobilewhep.client.utils.PeerConnectionFactoryHelper
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Request
import okhttp3.Response
import org.webrtc.AudioSource
import org.webrtc.Camera1Enumerator
import org.webrtc.Camera2Enumerator
import org.webrtc.CameraEnumerationAndroid
import org.webrtc.CameraEnumerator
import org.webrtc.CameraVideoCapturer
import org.webrtc.MediaConstraints
import org.webrtc.MediaStreamTrack
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpTransceiver
import org.webrtc.ScreenCapturerAndroid
import org.webrtc.SessionDescription
import org.webrtc.Size
import org.webrtc.SurfaceTextureHelper
import org.webrtc.VideoSource
import org.webrtc.VideoTrack
import java.io.EOFException
import java.io.IOException
import java.net.ConnectException
import java.net.URI
import java.net.URL
import java.util.UUID
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.math.roundToInt

data class WhipConfigurationOptions(
  val audioEnabled: Boolean = true,
  val videoEnabled: Boolean = true,
  val stunServerUrl: String? = null,
  val videoParameters: VideoParameters = VideoParameters.presetHD169,
  val videoDevice: String? = null,
  val preferredVideoCodecs: List<String>,
  val preferredAudioCodecs: List<String>,
  val isScreenSharingMode: Boolean = false
)

class WhipClient(
  appContext: Context,
  private val configOptions: WhipConfigurationOptions
) : ClientBase(
    appContext,
    stunServerUrl = configOptions.stunServerUrl
  ) {
  override var videoTrack: VideoTrack? = null
  private var videoCapturer: CameraVideoCapturer? = null
  private var videoSource: VideoSource? = null
  private var audioSource: AudioSource? = null
  private var screenCapturer: ScreenCapturerAndroid? = null
  private var cameraSurfaceTextureHelper: SurfaceTextureHelper? = null
  private var screenSurfaceTextureHelper: SurfaceTextureHelper? = null

  public var currentCameraDeviceId: String? = null
  private val peerConnectionFactory = PeerConnectionFactoryHelper.getWhipFactory(appContext, eglBase)
  private var isSharingScreen: Boolean = false
  private val isScreenSharingMode: Boolean

  init {
    isScreenSharingMode = configOptions.isScreenSharingMode
    if (!isScreenSharingMode) {
      setUpVideoAndAudioDevices()
    }
  }

  fun getPeerConnectionFactory(): PeerConnectionFactory = peerConnectionFactory

  fun setupPeerConnection() {
    super.setupPeerConnection(peerConnectionFactory)

    val audioEnabled = configOptions.audioEnabled
    val videoEnabled = configOptions.videoEnabled
    val direction = RtpTransceiver.RtpTransceiverDirection.SEND_ONLY

    Log.d(
      CLIENT_TAG,
      "Setting up peer connection - videoTrack: ${videoTrack != null}, audioTrack: ${audioTrack != null}, isSharingScreen: $isSharingScreen"
    )

    if (videoEnabled && videoTrack != null) {
      val transceiverInit = RtpTransceiver.RtpTransceiverInit(direction)
      val transceiver = peerConnection?.addTransceiver(videoTrack, transceiverInit)
      Log.d(CLIENT_TAG, "Added video transceiver for ${if (isScreenSharingMode) "screen share" else "camera"}")
      setCodecPreferencesIfAvailable(
        transceiver,
        configOptions.preferredVideoCodecs,
        MediaStreamTrack.MediaType.MEDIA_TYPE_VIDEO,
        factory = peerConnectionFactory
      )
    } else if (videoEnabled) {
      Log.e(CLIENT_TAG, "Video enabled but videoTrack is null!")
    }

    if (audioEnabled && audioTrack != null) {
      val audioTransceiverInit = RtpTransceiver.RtpTransceiverInit(direction)
      val transceiver = peerConnection?.addTransceiver(audioTrack, audioTransceiverInit)
      Log.d(CLIENT_TAG, "Added audio transceiver")
      setCodecPreferencesIfAvailable(
        transceiver,
        configOptions.preferredAudioCodecs,
        MediaStreamTrack.MediaType.MEDIA_TYPE_AUDIO,
        factory = peerConnectionFactory
      )
    } else if (audioEnabled) {
      Log.e(CLIENT_TAG, "Audio enabled but audioTrack is null!")
    }
  }

  /**
   * Gets the video and audio devices, prepares them, starts capture and adds it to the Peer Connection.
   *
   * @throws CaptureDeviceError.VideoDeviceNotAvailable if there is no video device.
   */
  private fun setUpVideoAndAudioDevices() {
    val audioEnabled = configOptions.audioEnabled
    val videoEnabled = configOptions.videoEnabled

    if (videoEnabled && configOptions.videoDevice == null) {
      throw CaptureDeviceError.VideoDeviceNotAvailable("Video device not found. Check if it can be accessed and passed to the constructor.")
    }

    if (!audioEnabled && !videoEnabled) {
      Log.d(
        CLIENT_TAG,
        "Both audioEnabled and videoEnabled is set to false, which will result in no stream at all. " +
          "Consider changing one of the options to true."
      )
    }

    if (videoEnabled) {
      val videoTrackId = UUID.randomUUID().toString()

      val cameraEnumerator: CameraEnumerator =
        if (Camera2Enumerator.isSupported(appContext)) {
          Camera2Enumerator(appContext)
        } else {
          Camera1Enumerator(false)
        }

      val videoCapturer: CameraVideoCapturer? =
        configOptions.videoDevice.let {
          cameraEnumerator.createCapturer(it, null)
        }

      val videoSource: VideoSource =
        peerConnectionFactory.createVideoSource(videoCapturer!!.isScreencast)
      val surfaceTextureHelper = SurfaceTextureHelper.create("CaptureThread", eglBase.eglBaseContext)
      cameraSurfaceTextureHelper = surfaceTextureHelper
      videoCapturer.initialize(surfaceTextureHelper, appContext, videoSource.capturerObserver)
      val videoSize =
        setVideoSize(
          cameraEnumerator,
          configOptions.videoDevice,
          configOptions.videoParameters
        )
      try {
        videoCapturer.startCapture(
          videoSize!!.width,
          videoSize.height,
          configOptions.videoParameters.maxFps
        )
      } catch (e: Exception) {
        throw CaptureDeviceError.VideoSizeNotSupported(
          "VideoSize ${configOptions.videoParameters} is not supported by this device. Consider switching to another preset."
        )
      }

      val videoTrack: VideoTrack = peerConnectionFactory.createVideoTrack(videoTrackId, videoSource)

      this.videoSource = videoSource
      this.videoCapturer = videoCapturer
      this.currentCameraDeviceId = configOptions.videoDevice

      videoTrack.setEnabled(true)
      this.videoTrack = videoTrack
    }

    if (audioEnabled) {
      val audioTrackId = UUID.randomUUID().toString()
      val audioSource = this.peerConnectionFactory.createAudioSource(MediaConstraints())
      val audioTrack = this.peerConnectionFactory.createAudioTrack(audioTrackId, audioSource)

      this.audioSource = audioSource
      this.audioTrack = audioTrack
    }

    peerConnection?.enforceSendOnlyDirection()
  }

  private val REQUIRED_PERMISSIONS =
    arrayOf(Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO)

  /**
   * Connects the client to the WHIP server using WebRTC Peer Connection.
   *
   * @throws SessionNetworkError.ConfigurationError if the stunServerUrl parameter
   *  of the initial configuration is incorrect, which leads to peerConnection being nil
   *  or in any other case where there has been an error in creating the peerConnection
   *
   */
  override suspend fun connect(connectOptions: ClientConnectOptions) {
    try {
      super.connect(connectOptions)

      if (isScreenSharingMode) {
        if (videoTrack == null ||
          screenCapturer == null ||
          (configOptions.audioEnabled && audioTrack == null)
        ) {
          Log.w(CLIENT_TAG, "Screen sharing setup incomplete during connect.\nAudio track: $audioTrack,\nVideo track: $videoTrack")
          throw SessionNetworkError.ConfigurationError(
            "Failed to connect: screen sharing setup not complete. Check if permissions were granted."
          )
        }
      } else {
        if (videoTrack == null ||
          videoCapturer == null ||
          (configOptions.audioEnabled && audioTrack == null)
        ) {
          setUpVideoAndAudioDevices()
        }
      }

      if (peerConnection == null) {
        setupPeerConnection()
      }
      val requiredPermissions =
        if (isScreenSharingMode) {
          if (configOptions.audioEnabled) {
            arrayOf(Manifest.permission.RECORD_AUDIO)
          } else {
            emptyArray()
          }
        } else {
          REQUIRED_PERMISSIONS
        }

      if (requiredPermissions.isNotEmpty() && !hasPermissions(appContext, requiredPermissions)) {
        val permissionType = if (isScreenSharingMode) "audio recording" else "camera and audio recording"
        throw PermissionError.PermissionsNotGrantedError(
          "Permissions for $permissionType have not been granted. Please check your application settings."
        )
      }

      val constraints = MediaConstraints()
      peerConnection?.let {
        Log.d(CLIENT_TAG, "Creating SDP offer for ${if (isScreenSharingMode) "screen share" else "camera"}")
        val sdpOffer = it.createOffer(constraints).getOrThrow()
        it.setLocalDescription(sdpOffer).getOrThrow()
        val sdp = sendSdpOffer(sdpOffer.description)
        iceCandidates.forEach { sendCandidate(it) }

        val answer =
          SessionDescription(
            SessionDescription.Type.ANSWER,
            sdp
          )
        it.setRemoteDescription(answer)
      } ?: {
        throw SessionNetworkError.ConfigurationError("Failed to connect: no peer connection")
      }
    } catch (e: PermissionError.PermissionsNotGrantedError) {
      cleanupPeerConnection()
      throw e
    } catch (e: SessionNetworkError) {
      cleanupPeerConnection()
      throw e
    } catch (e: AttributeNotFoundError) {
      cleanupPeerConnection()
      throw e
    } catch (e: Exception) {
      Log.e(CLIENT_TAG, "Failed to connect: ${e.message}", e)
      cleanupPeerConnection()
      throw SessionNetworkError.ConnectionError("Connection failed: ${e.message ?: e.javaClass.simpleName}")
    }
  }

  fun stopScreenShare() {
    if (!isSharingScreen) {
      Log.w(CLIENT_TAG, "Screen sharing is not active")
      return
    }

    Log.d(CLIENT_TAG, "Stopping screen share")

    cleanupScreenCapturer()
    cleanupVideoTrack()

    isSharingScreen = false

    Log.d(CLIENT_TAG, "Screen share stopped")
  }

  /**
   * Closes the established Peer Connection.
   *
   * @throws SessionNetworkError.ConfigurationError if the stunServerUrl parameter
   *  of the initial configuration is incorrect, which leads to peerConnection being nil
   *  or in any other case where there has been an error in creating the peerConnection
   *
   */
  suspend fun disconnect() {
    if (isSharingScreen) {
      stopScreenShare()
    }
    cleanupPeerConnection()
    disconnectResource()
  }

  private fun cleanupPeerConnection() {
    peerConnection?.close()
    peerConnection?.dispose()
    peerConnection = null
  }

  private fun cleanupVideoCapturer() {
    videoCapturer?.stopCapture()
    videoCapturer?.dispose()
    videoCapturer = null

    cameraSurfaceTextureHelper?.dispose()
    cameraSurfaceTextureHelper = null
  }

  private fun cleanupScreenCapturer() {
    screenCapturer?.stopCapture()
    screenCapturer?.dispose()
    screenCapturer = null

    screenSurfaceTextureHelper?.dispose()
    screenSurfaceTextureHelper = null
  }

  private fun cleanupVideoTrack() {
    videoSource?.dispose()
    videoSource = null

    videoTrack?.dispose()
    videoTrack = null
  }

  private fun cleanupAudioTrack() {
    audioSource?.dispose()
    audioSource = null

    audioTrack?.dispose()
    audioTrack = null
  }

  fun cleanup() {
    cleanupVideoCapturer()
    cleanupScreenCapturer()
    cleanupVideoTrack()
    cleanupAudioTrack()
    cleanupPeerConnection()
    cleanupFactory()
    cleanupEglBase()
  }

  fun switchCamera(deviceId: String) {
    if (isSharingScreen) {
      Log.w(CLIENT_TAG, "Cannot switch camera while screen sharing")
      return
    }

    val enumerator: CameraEnumerator =
      if (Camera2Enumerator.isSupported(appContext)) Camera2Enumerator(appContext) else Camera1Enumerator(false)

    val availableDevices = enumerator.deviceNames
    if (!availableDevices.contains(deviceId)) {
      Log.w(CLIENT_TAG, "Device with ID $deviceId not found. Available devices: $availableDevices")
      return
    }

    try {
      videoCapturer?.switchCamera(
        object : CameraVideoCapturer.CameraSwitchHandler {
          override fun onCameraSwitchDone(isFrontCamera: Boolean) {
            // Camera switch completed successfully
            currentCameraDeviceId = deviceId
          }

          override fun onCameraSwitchError(errorDescription: String?) {
            Log.e(CLIENT_TAG, "Camera switch error: $errorDescription")
          }
        },
        deviceId
      )
    } catch (e: Exception) {
      Log.e(CLIENT_TAG, "Failed to switch camera to $deviceId: ${e.message}")
    }
  }

  /**
   * Starts screen sharing using MediaProjection.
   * Note: MediaProjection permission must be obtained at the Activity level before calling this.
   *
   * @param mediaProjectionIntent The intent received from MediaProjection permission dialog
   */
  fun startScreenShare(mediaProjectionIntent: Intent) {
    Log.d(CLIENT_TAG, "Starting screen share with MediaProjection")

    isSharingScreen = true

    cleanupVideoCapturer()
    cleanupVideoTrack()

    val screenCapturer =
      ScreenCapturerAndroid(
        mediaProjectionIntent,
        object : android.media.projection.MediaProjection.Callback() {
          override fun onStop() {
            super.onStop()
            Log.d(CLIENT_TAG, "Screen capture stopped")
            isSharingScreen = false
          }
        }
      )

    val videoSource: VideoSource = peerConnectionFactory.createVideoSource(true)
    val surfaceTextureHelper = SurfaceTextureHelper.create("ScreenCaptureThread", eglBase.eglBaseContext)
    screenSurfaceTextureHelper = surfaceTextureHelper

    screenCapturer.initialize(surfaceTextureHelper, appContext, videoSource.capturerObserver)

    val screenDimensions = getScreenDimensions()
    val targetDimensions = downscaleResolution(screenDimensions, configOptions.videoParameters.dimensions)

    try {
      screenCapturer.startCapture(
        targetDimensions.width,
        targetDimensions.height,
        configOptions.videoParameters.maxFps
      )
      Log.d(CLIENT_TAG, "Screen capture started with dimensions ${targetDimensions.width}x${targetDimensions.height}")
    } catch (e: Exception) {
      Log.e(CLIENT_TAG, "Failed to start screen capture: ${e.message}", e)
      throw CaptureDeviceError.VideoSizeNotSupported(
        "Failed to start screen capture: ${e.message}"
      )
    }

    val videoTrackId = UUID.randomUUID().toString()
    val videoTrack: VideoTrack = peerConnectionFactory.createVideoTrack(videoTrackId, videoSource)

    this.videoSource = videoSource
    this.screenCapturer = screenCapturer
    this.currentCameraDeviceId = null

    videoTrack.setEnabled(true)
    this.videoTrack = videoTrack

    if (configOptions.audioEnabled && audioTrack == null) {
      val audioTrackId = UUID.randomUUID().toString()
      val audioSource = peerConnectionFactory.createAudioSource(MediaConstraints())
      val audioTrack = peerConnectionFactory.createAudioTrack(audioTrackId, audioSource)
      audioTrack.setEnabled(true)
      this.audioSource = audioSource
      this.audioTrack = audioTrack
      Log.d(CLIENT_TAG, "Audio track created for screen share")
    }

    notifyTrackListeners()

    Log.d(CLIENT_TAG, "Screen share setup complete")
  }

  protected fun cleanupFactory() {
    PeerConnectionFactoryHelper.clearWhipFactory()
  }

  private suspend fun disconnectResource() {
    suspendCancellableCoroutine { continuation ->
      if (connectOptions == null) {
        continuation.resumeWithException(
          SessionNetworkError.ConnectionError(
            "Cannot DELETE. Connection not setup. Remember to call connect first."
          )
        )
        return@suspendCancellableCoroutine
      }

      val serverUrl = URL(connectOptions!!.serverUrl)

      val requestURL =
        URI(serverUrl.protocol, null, serverUrl.host, serverUrl.port, patchEndpoint, null, null)

      var requestBuilder: Request.Builder =
        Request
          .Builder()
          .url(requestURL.toURL())
          .delete()
      if (connectOptions!!.authToken != null) {
        requestBuilder =
          requestBuilder.header("Authorization", "Bearer " + connectOptions!!.authToken)
      }

      val request = requestBuilder.build()
      val requestCall = client.newCall(request)

      continuation.invokeOnCancellation {
        requestCall.cancel()
      }

      requestCall.enqueue(
        object : Callback {
          override fun onFailure(
            call: Call,
            e: IOException
          ) {
            val errorMessage =
              when {
                e is ConnectException -> {
                  "DELETE Failed, network error. Check if the server is up and running and the token and the server url is correct."
                }
                e is EOFException || e.message?.contains("unexpected end of stream") == true -> {
                  "Server closed connection unexpectedly during disconnect. The WHIP server may have already terminated the session or crashed."
                }
                else -> {
                  "Failed to disconnect: ${e.message}"
                }
              }

            Log.e(CLIENT_TAG, "Disconnect failed: $errorMessage", e)
            continuation.resumeWithException(
              SessionNetworkError.ConnectionError(errorMessage)
            )
          }

          override fun onResponse(
            call: Call,
            response: Response
          ) {
            response.use {
              if (!response.isSuccessful) {
                val exception =
                  AttributeNotFoundError.ResponseNotFound(
                    "DELETE Failed, invalid response. Status code: ${response.code}. Check if the server is up and running and the token and the server url is correct."
                  )
                continuation.resumeWithException(exception)
              } else {
                continuation.resume(Unit)
              }
            }
          }
        }
      )
    }
  }

  private fun PeerConnection.enforceSendOnlyDirection() {
    transceivers.forEach { transceiver ->
      if (transceiver.direction == RtpTransceiver.RtpTransceiverDirection.SEND_RECV) {
        transceiver.direction = RtpTransceiver.RtpTransceiverDirection.SEND_ONLY
      }
    }
  }

  private fun setVideoSize(
    enumerator: CameraEnumerator,
    deviceName: String?,
    videoParameters: VideoParameters
  ): Size? {
    val sizes =
      enumerator
        .getSupportedFormats(deviceName)
        ?.map { Size(it.width, it.height) }
        ?: emptyList()

    val size =
      CameraEnumerationAndroid.getClosestSupportedSize(
        sizes,
        videoParameters.dimensions.width,
        videoParameters.dimensions.height
      )

    return size
  }

  private fun getScreenDimensions(): Dimensions {
    val windowManager = appContext.getSystemService(Context.WINDOW_SERVICE) as WindowManager
    val displayMetrics = DisplayMetrics()
    windowManager.defaultDisplay.getRealMetrics(displayMetrics)

    return Dimensions(width = displayMetrics.widthPixels, height = displayMetrics.heightPixels)
  }

  private fun downscaleResolution(
    from: Dimensions,
    to: Dimensions
  ): Dimensions =
    when {
      from.height > to.height -> {
        val ratio = from.height.toFloat() / from.width.toFloat()
        val newHeight = to.height
        val newWidth = (newHeight.toFloat() / ratio).roundToInt()
        Dimensions(width = newWidth, height = newHeight)
      }
      from.width > to.width -> {
        val ratio = from.height.toFloat() / from.width.toFloat()
        val newWidth = to.width
        val newHeight = (newWidth.toFloat() * ratio).roundToInt()
        Dimensions(width = newWidth, height = newHeight)
      }
      else -> from
    }

  private fun hasPermissions(
    context: Context,
    permissions: Array<String>
  ): Boolean {
    for (permission in permissions) {
      if (ContextCompat.checkSelfPermission(
          context,
          permission
        ) != PackageManager.PERMISSION_GRANTED
      ) {
        return false
      }
    }
    return true
  }

  companion object {
    private fun getEnumerator(context: Context): CameraEnumerator =
      if (Camera2Enumerator.isSupported(context)) {
        Camera2Enumerator(context)
      } else {
        Camera1Enumerator(true)
      }

    fun getCaptureDevices(context: Context): List<CaptureDevice> {
      val enumerator = getEnumerator(context)
      return enumerator.deviceNames.map { name ->
        CaptureDevice(
          name,
          enumerator.isFrontFacing(name),
          enumerator.isBackFacing(name)
        )
      }
    }
  }
}

File: packages/ios-client/Sources/ios-client/WhipClient.swift
import WebRTC
import ReplayKit
import os

public struct WhipConfigurationOptions {
    public let audioEnabled: Bool
    public let videoEnabled: Bool
    public let videoParameters: VideoParameters
    public let stunServerUrl: String?
    public let preferredVideoCodecs: [String]
    public let preferredAudioCodecs: [String]

    public init(
        audioEnabled: Bool = true, videoEnabled: Bool = true,
        videoParameters: VideoParameters, stunServerUrl: String?, preferredVideoCodecs: [String] = [],
        preferredAudioCodecs: [String] = []
    ) {
        self.audioEnabled = audioEnabled
        self.videoEnabled = videoEnabled
        self.videoParameters = videoParameters
        self.stunServerUrl = stunServerUrl
        self.preferredVideoCodecs = preferredVideoCodecs
        self.preferredAudioCodecs = preferredAudioCodecs
    }
}

public class WhipClient: ClientBase {
    private var configOptions: WhipConfigurationOptions
    private var videoCapturer: RTCCameraVideoCapturer?
    private var videoSource: RTCVideoSource?
    
    private var broadcastScreenShareReceiver: BroadcastScreenShareReceiver?
    private var broadcastScreenShareCapturer: BroadcastScreenShareCapturer?

    public var currentCameraDeviceId: String?
    public var isScreenShareOn: Bool = false

    /**
    Initializes a `WhipClient` object.
    - Parameter configurationOptions: Additional configuration options, such as a STUN server URL or authorization token.
    
    - Returns: A `WhipClient` object.
    */
    public init(
        configOptions: WhipConfigurationOptions
    ) {
        self.configOptions = configOptions
        super.init(stunServerUrl: configOptions.stunServerUrl)

        do {
            try setUpVideoAndAudioDevices()
        } catch let error as CaptureDeviceError {
            switch error {
            case .VideoDeviceNotAvailable(let description),
                .VideoSizeNotSupported(let description):
                print(description)
            }
        } catch {
            print("Unexpected error: \(error)")
        }
    }

    deinit {
        cleanup()
    }

    override func setUpPeerConnection() throws {
        try super.setUpPeerConnection()

        let audioEnabled = configOptions.audioEnabled
        let videoEnabled = configOptions.videoEnabled

        if !audioEnabled && !videoEnabled {
            logger.warning(
                "Both audioEnabled and videoEnabled are set to false, what will result in no stream at all. Consider changing one of the options to true."
            )
        }

        let sendEncodings = [RTCRtpEncodingParameters.create(active: true)]
        let localStreamId = UUID().uuidString

        if let videoTrack = self.videoTrack, videoEnabled {
            let transceiverInit = RTCRtpTransceiverInit()
            transceiverInit.direction = RTCRtpTransceiverDirection.sendOnly
            transceiverInit.streamIds = [localStreamId]
            transceiverInit.sendEncodings = sendEncodings

            let transceiver = peerConnection?.addTransceiver(with: videoTrack, init: transceiverInit)
            setCodecPreferencesIfAvailable(
                transceiver: transceiver,
                preferredCodecs: configOptions.preferredVideoCodecs,
                mediaType: kRTCMediaStreamTrackKindVideo
            )
        }

        if let audioTrack = self.audioTrack, audioEnabled {
            let audioTransceiverInit = RTCRtpTransceiverInit()
            audioTransceiverInit.direction = RTCRtpTransceiverDirection.sendOnly
            audioTransceiverInit.streamIds = [localStreamId]
            let transceiver = peerConnection?.addTransceiver(with: audioTrack, init: audioTransceiverInit)
            peerConnection?.enforceSendOnlyDirection()
            setCodecPreferencesIfAvailable(
                transceiver: transceiver,
                preferredCodecs: configOptions.preferredAudioCodecs,
                mediaType: kRTCMediaStreamTrackKindAudio
            )

            let audioSession = RTCAudioSession.sharedInstance()
            audioSession.lockForConfiguration()
            let config = RTCAudioSessionConfiguration.webRTC()
            config.category = AVAudioSession.Category.playAndRecord.rawValue
            config.mode = videoEnabled ? AVAudioSession.Mode.videoChat.rawValue : AVAudioSession.Mode.voiceChat.rawValue
            try audioSession.setConfiguration(config, active: true)
            audioSession.unlockForConfiguration()
        }
    }

    /**
    Connects the client to the WHIP server using WebRTC Peer Connection.
    
    - Throws: `SessionNetworkError.ConfigurationError` if the `stunServerUrl` parameter
        of the initial configuration is incorrect, which leads to `peerConnection` being nil or in any other case where there has been an error in creating the `peerConnection`
    */
    public override func connect(_ connectOptions: ClientConnectOptions) async throws {
        try await super.connect(connectOptions)
        if !self.isConnectionSetUp {
            try setUpPeerConnection()
        } else if self.isConnectionSetUp && self.peerConnection == nil {
            throw SessionNetworkError.ConfigurationError(
                description: "Failed to establish RTCPeerConnection. Check initial configuration")
        }

        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        let offer = try await peerConnection!.offer(for: constraints)
        try await peerConnection!.setLocalDescription(offer)
        let sdpAnswer = try await send(sdpOffer: offer.sdp)

        for candidate in iceCandidates {
            try await sendCandidate(candidate: candidate)
        }
        let remoteDescription = RTCSessionDescription(type: .answer, sdp: sdpAnswer)

        try await peerConnection!.setRemoteDescription(remoteDescription)
    }

    /**
    Closes the established Peer Connection.
    
    - Throws: `SessionNetworkError.ConfigurationError` if the `stunServerUrl` parameter
    of the initial configuration is incorrect, which leads to `peerConnection` being nil or in any other case where there has been an error in creating the `peerConnection`
    */
    public func disconnect() async throws {
        DispatchQueue.main.sync { [weak self] in
            self?.peerConnection?.close()
            self?.peerConnection = nil
            self?.isConnectionSetUp = false
        }
        try await disconnectResource()
    }

    public func cleanup() {
        peerConnection?.close()
        videoCapturer?.stopCapture()
    }

    public func disconnectResource() async throws {
        guard let connectOptions else {
            throw SessionNetworkError.ConnectionError(
                description:
                    "Connection not setup. Remember to call connect first.")
        }
        guard let patchEndpoint = self.patchEndpoint else {
            throw AttributeNotFoundError.PatchEndpointNotFound(
                description: "Patch endpoint not found. Make sure the SDP answer is correct.")
        }
        var components = URLComponents(string: connectOptions.serverUrl.absoluteString)
        components?.path = patchEndpoint

        let url = components?.url
        var request = URLRequest(url: url!)
        request.httpMethod = "DELETE"

        if let token = connectOptions.authToken {
            request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let response: URLResponse
        (_, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            return
        }
        if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 {
            throw SessionNetworkError.ConnectionError(
                description:
                    "DELETE Failed, invalid response. Check if the server is up and running and the token and the server url is correct."
            )
        }
    }

    public func switchCamera(deviceId: String) {
        guard let videoCapturer else {
            logger.error("Switch camera failed, videoCapturer is nil")
            return
        }
        videoCapturer.stopCapture()

        let devices = RTCCameraVideoCapturer.captureDevices()

        if let device = devices.first(where: { $0.uniqueID == deviceId }) {
            startCapture(device)
        } else {
            print("Device with ID \(deviceId) not found")
            return
        }
    }

    private func setUpVideoAndAudioDevices() throws {
        let audioEnabled = configOptions.audioEnabled
        let videoEnabled = configOptions.videoEnabled

        if !audioEnabled && !videoEnabled {
            logger.warning(
                "Both audioEnabled and videoEnabled are set to false, what will result in no stream at all. Consider changing one of the options to true."
            )
        }

        if videoEnabled {
            let videoSource = WhipClient.peerConnectionFactory.videoSource()
            self.videoSource = videoSource
            let videoCapturer = RTCCameraVideoCapturer(delegate: videoSource, captureSession: AVCaptureSession())
            self.videoCapturer = videoCapturer
            let videoTrackId = UUID().uuidString

            let videoTrack = WhipClient.peerConnectionFactory.videoTrack(with: videoSource, trackId: videoTrackId)
            videoTrack.isEnabled = true

            self.videoTrack = videoTrack
        }
        
        if audioEnabled {
            let audioTrackId = UUID().uuidString
            let audioSource = WhipClient.peerConnectionFactory.audioSource(with: nil)
            let audioTrack = WhipClient.peerConnectionFactory.audioTrack(with: audioSource, trackId: audioTrackId)

            self.audioTrack = audioTrack
        }

    }

    public func startCapture(_ videoDevice: AVCaptureDevice) {
        let videoParameters = configOptions.videoParameters

        let (format, fps) = setVideoSize(
            device: videoDevice, videoParameters: (videoParameters))

        guard let videoCapturer else {
            logger.warning("No video capturer to start")
            return
        }

        videoCapturer.startCapture(with: videoDevice, format: format, fps: fps) { [weak self] error in
            if let error = error {
                print("Error starting the video capture: \(error)")
            } else {
                print("Video capturing started")
                self?.currentCameraDeviceId = videoDevice.uniqueID
            }
        }
    }

    private func setVideoSize(device: AVCaptureDevice, videoParameters: VideoParameters) -> (
        selectedFormat: AVCaptureDevice.Format, fps: Int
    ) {
        let formats: [AVCaptureDevice.Format] = RTCCameraVideoCapturer.supportedFormats(for: device)

        let (targetWidth, targetHeight) = (
            videoParameters.dimensions.width,
            videoParameters.dimensions.height
        )

        var currentDiff = Int32.max
        var selectedFormat: AVCaptureDevice.Format = formats[0]
        for format in formats {
            let dimension = CMVideoFormatDescriptionGetDimensions(format.formatDescription)

            let diff = abs(targetWidth - dimension.width) + abs(targetHeight - dimension.height)
            if diff < currentDiff {
                selectedFormat = format
                currentDiff = diff
            }
        }

        let fps = videoParameters.maxFps

        // discover FPS limits
        var minFps = 60
        var maxFps = 0
        for fpsRange in selectedFormat.videoSupportedFrameRateRanges {
            minFps = min(minFps, Int(fpsRange.minFrameRate))
            maxFps = max(maxFps, Int(fpsRange.maxFrameRate))
        }
        if fps < minFps || fps > maxFps {
            fatalError("unsported requested frame rate of (\(minFps) - \(maxFps)")
        }

        return (selectedFormat, fps)
    }

    public static func getCaptureDevices() -> [AVCaptureDevice] {
        return RTCCameraVideoCapturer.captureDevices()
    }

    // MARK: - Codec Management

    /**
     Gets the names of supported sender video codecs.
    
     - Returns: Array of supported video codec names
     */
    public static func getSupportedSenderVideoCodecsNames() -> [String] {
        let capabilities = WhipClient.peerConnectionFactory.rtpSenderCapabilities(
            forKind: kRTCMediaStreamTrackKindVideo)

        return capabilities.codecs.map { $0.name }
    }

    /**
     Gets the names of supported sender audio codecs.
    
     - Returns: Array of supported audio codec names
     */
    public static func getSupportedSenderAudioCodecsNames() -> [String] {
        let capabilities = WhipClient.peerConnectionFactory.rtpSenderCapabilities(
            forKind: kRTCMediaStreamTrackKindAudio)

        return capabilities.codecs.map { $0.name }
    }

    /**
     Sets preferred video codecs for sending.
    
     - Parameter preferredCodecs: Array of preferred video codec names, or nil to skip setting
     */
    public func setPreferredVideoCodecs(preferredCodecs: [String]?) {
        guard let preferredCodecs = preferredCodecs, !preferredCodecs.isEmpty else {
            return
        }

        guard let peerConnection = peerConnection else {
            return
        }

        for transceiver in peerConnection.transceivers {
            if transceiver.mediaType == .video {
                setCodecPreferencesIfAvailable(
                    transceiver: transceiver,
                    preferredCodecs: preferredCodecs,
                    mediaType: kRTCMediaStreamTrackKindVideo
                )
            }
        }
    }

    /**
     Sets preferred audio codecs for sending.
    
     - Parameter preferredCodecs: Array of preferred audio codec names, or nil to skip setting
     */
    public func setPreferredAudioCodecs(preferredCodecs: [String]?) {
        guard let preferredCodecs = preferredCodecs, !preferredCodecs.isEmpty else {
            return
        }

        guard let peerConnection = peerConnection else {
            return
        }

        for transceiver in peerConnection.transceivers {
            if transceiver.mediaType == .audio {
                setCodecPreferencesIfAvailable(
                    transceiver: transceiver,
                    preferredCodecs: preferredCodecs,
                    mediaType: kRTCMediaStreamTrackKindAudio
                )
            }
        }
    }

    // MARK: - Screen Sharing

    /**
     Starts screen sharing mode.
     
     This method:
     - Stops camera capture
     - Creates a new video source for screen sharing
     - Starts IPC server listening for frames from the broadcast extension
     - Shows the system broadcast picker for the user to select the extension
     
     This should be called during initialization, not during an active stream.
     */
    public func startScreenShare() throws {
        guard let screenShareExtensionBundleId = Bundle.main.infoDictionary?["ScreenShareExtensionBundleId"] as? String else {
            throw ScreenSharingError.NoExtension(description: "No screen share extension bundle id set. Please set ScreenShareExtensionBundleId in Info.plist")
        }
        
        let appGroup = Bundle.main.object(forInfoDictionaryKey: "AppGroupName") as? String
            ?? "group.com.swmansion.mobilewhepclient"
        
        videoCapturer?.stopCapture()
        
        let videoSource = WhipClient.peerConnectionFactory.videoSource(forScreenCast: true)
        self.videoSource = videoSource
        
        broadcastScreenShareReceiver = BroadcastScreenShareReceiver(
            onStart: { [weak self] in
                self?.logger.info("Screen broadcast started")
            },
            onStop: { [weak self] in
                self?.logger.info("Screen broadcast stopped")
            }
        )
        
        broadcastScreenShareCapturer = BroadcastScreenShareCapturer(
            videoSource,
            appGroup: appGroup,
            videoParameters: .presetFHD169,
            delegate: broadcastScreenShareReceiver
        )
        
        broadcastScreenShareCapturer?.startListening()
        isScreenShareOn = true
        
        let videoTrackId = UUID().uuidString
        let videoTrack = WhipClient.peerConnectionFactory.videoTrack(with: videoSource, trackId: videoTrackId)
        videoTrack.isEnabled = true
        self.videoTrack = videoTrack
        
        logger.info("Screen sharing initialized, showing broadcast picker")
        
        DispatchQueue.main.async {
            RPSystemBroadcastPickerView.show(for: screenShareExtensionBundleId)
        }
    }
}
